File: 104. Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque

class Solution(object):
    
    # BFS: O(N) time & space
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root: return 0
        queue, depth = deque([root]), 0
        while queue:
            depth, size = depth+1, len(queue)
            while size:
                node, size = queue.popleft(), size-1
                queue += filter(None, [node.left, node.right])
        return depth
        
    # DFS: O(N) time & space
    def maxDepth(self, root):
        return self.dfs(root, 0)
    
    def dfs(self, root, depth):
        if not root: return depth
        leftDepth = self.dfs(root.left, depth+1)
        rightDepth = self.dfs(root.right, depth+1)
        return max(leftDepth, rightDepth)
    
    # DFS Pythonic: O(N) time & space
    def maxDepth(self, root):
        if not root: return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))
        
        
        
            

File: 104. Maximum Depth of Binary Tree.cpp
#include <iostream>
#include <queue>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // DFS with elegancy: O(Node), 16.40%, 6ms
    int maxDepth(TreeNode* root) {
        if(root == NULL) return 0;
        else return max(maxDepth(root->left), maxDepth(root->right)) + 1;
    }

    // BFS: 16ms
    int maxDepth(TreeNode* root){
        if(root == NULL) return 0;
        queue<TreeNode *> que;
        que.push(root);
        int depth = 0;
        while(!que.empty()){
            depth ++;       // every layer of tree nodes
            // bug: i<que.size(), size of queue change from time to time, should settle down the value before each for iteration.
            int size = que.size();
            for(int i = 0; i < size; i ++){
                TreeNode *temp = que.front();
                que.pop();
                if(temp->left != NULL) que.push(temp->left);
                if(temp->right != NULL) que.push(temp->right);
            }
        }
        return depth;
    }
};
